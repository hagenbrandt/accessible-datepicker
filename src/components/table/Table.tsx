import React from 'react';

type TableType = {
  head: string;
  headerRowTitle: string;
  bodyTitle: string;
  bodyRowTitle: string;
  cell: string;
};

type TableRowType = {
  tableRowTitle: string;
  cellItems: string[];
  rowType: CellType;
  rowKey?: string;
};

type TableCellType = {
  data: string;
  cellType: CellType;
  cellKey?: number;
};

export type CellType = 'headerCell' | 'dataCell';

export const Table = ({ head, headerRowTitle, bodyTitle, bodyRowTitle, cell }: TableType) => {
  return (
    <table>
      <thead title={headerRowTitle}>
        <TableRow
          tableRowTitle={headerRowTitle}
          cellItems={[head]}
          rowType="headerCell"
          rowKey="header-row-1"
        />
      </thead>
      <tbody title={bodyTitle}>
        <TableRow
          tableRowTitle={bodyRowTitle}
          cellItems={[cell]}
          rowType="dataCell"
          rowKey="body-row-key-1"
        />
      </tbody>
    </table>
  );
};

export const TableCell = ({ data, cellType, cellKey }: TableCellType) => {
  if (!data) {
    return <></>;
  }

  const keyForCell = cellKey ? `${data}-${cellKey}` : data;

  if (cellType === 'headerCell') {
    return <th key={keyForCell}>{data}</th>;
  }

  return <td key={keyForCell}>{data}</td>;
};

export const TableRow = ({ tableRowTitle, cellItems, rowType, rowKey }: TableRowType) => {
  if (!cellItems.length) {
    return <></>;
  }

  const keyForRow = rowKey ? `${rowType}-${rowKey}` : rowType;

  return (
    <tr title={tableRowTitle} key={keyForRow}>
      {cellItems.map((item, index) => {
        return <TableCell data={item} cellType={rowType} key={index} />;
      })}
    </tr>
  );
};
