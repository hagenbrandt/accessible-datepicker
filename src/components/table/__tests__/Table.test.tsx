import React from 'react';
import { render, screen } from '@testing-library/react';
import { renderTableCell, renderTableRow, Table } from '../Table';

describe('table', () => {
  const head = 'Head 1';
  const headerRowTitle = 'Header';
  const bodyTitle = 'Body Title';
  const bodyRowTitle = 'Body Row 1';
  const cell = 'Table Element 1';
  const component = (
    <Table
      head={head}
      headerRowTitle={headerRowTitle}
      bodyTitle={bodyTitle}
      bodyRowTitle={bodyRowTitle}
      cell={cell}
    />
  );

  beforeEach(() => {
    render(component);
  });

  it('renders a table', () => {
    expect(screen.getByRole('table')).toBeInTheDocument();
  });

  it('renders a columnheader with a given string in a rowgroup', () => {
    expect(screen.getByRole('rowgroup', { name: headerRowTitle })).toBeInTheDocument();
    expect(screen.getByRole('columnheader')).toHaveTextContent(head);
    expect(screen.getByRole('row', { name: headerRowTitle })).toHaveTextContent(head);
  });

  it('renders a cell within a row and a tablebody', () => {
    expect(screen.getByRole('rowgroup', { name: bodyTitle })).toBeInTheDocument();
    expect(screen.getByRole('row', { name: bodyRowTitle })).toBeInTheDocument();
    expect(screen.getByRole('cell', { name: cell })).toBeInTheDocument();
  });
});

describe('renderTableCell', () => {
  const cellContent = 'Cell Content 1';
  const tableWrapperWithHeaderCell = (
    <table>
      <thead>
        <tr>{renderTableCell(cellContent, 'headerCell', 1)}</tr>
      </thead>
    </table>
  );
  const tableWrapperWithDataCell = (
    <table>
      <tbody>
        <tr>{renderTableCell(cellContent, 'dataCell', 2)}</tr>
      </tbody>
    </table>
  );

  it('renders a columnheader with given string', () => {
    render(tableWrapperWithHeaderCell);

    expect(screen.getByRole('columnheader')).toHaveTextContent(cellContent);
  });

  it('renders a data cell with given content', () => {
    render(tableWrapperWithDataCell);

    expect(screen.getByRole('cell')).toHaveTextContent(cellContent);
  });

  it('returns empty tag when no data is given', () => {
    const { container } = render(renderTableCell('', 'headerCell'));

    expect(container).toBeEmptyDOMElement();
  });
});

describe('renderTableRow', () => {
  const tableRowTitle = 'Table Row Title';
  const cellItems = ['Item 1', 'Item 2', 'Item 3'];
  const headerRowKey = 'Head-Key-1';
  const dataRowKey = 'Data-Key-1';

  const tableRowWithHeaderCells = (
    <table>
      <thead>{renderTableRow(tableRowTitle, cellItems, 'headerCell', headerRowKey)}</thead>
    </table>
  );
  const tableRowWithDataCells = (
    <table>
      <thead>{renderTableRow(tableRowTitle, cellItems, 'dataCell', dataRowKey)}</thead>
    </table>
  );

  it('renders a table row with header cells', () => {
    render(tableRowWithHeaderCells);

    expect(screen.getAllByRole('columnheader')).toHaveLength(cellItems.length);
    cellItems.forEach((item, index) => {
      expect(screen.getByRole('columnheader', { name: item })).toHaveTextContent(cellItems[index]);
    });
  });

  it('renders a table row with data cells', () => {
    render(tableRowWithDataCells);

    expect(screen.getAllByRole('cell')).toHaveLength(cellItems.length);
    cellItems.forEach((item, index) => {
      expect(screen.getByRole('cell', { name: item })).toHaveTextContent(cellItems[index]);
    });
  });

  it('table row has a title with given content', () => {
    render(tableRowWithHeaderCells);

    expect(screen.getByRole('row')).toHaveAttribute('title', tableRowTitle);
  });

  it('returns empty tag when no data is given', () => {
    const { container } = render(
      renderTableRow(tableRowTitle, [], 'dataCell', 'Key-For-Empty-Row'),
    );
    expect(container).toBeEmptyDOMElement();
  });
});
