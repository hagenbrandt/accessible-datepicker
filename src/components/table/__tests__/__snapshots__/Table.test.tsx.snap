// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderTableCell matches snapshot 1`] = `
<th>
  Cell Content 1
</th>
`;

exports[`renderTableCell matches snapshot 2`] = `
<td>
  Cell Content 1
</td>
`;

exports[`renderTableRow matches snapshot 1`] = `
<table>
  <thead>
    <tr
      title="Table Row Title"
    >
      <th>
        Item 1
      </th>
      <th>
        Item 2
      </th>
      <th>
        Item 3
      </th>
    </tr>
  </thead>
</table>
`;

exports[`renderTableRow matches snapshot 2`] = `
<table>
  <thead>
    <tr
      title="Table Row Title"
    >
      <td>
        Item 1
      </td>
      <td>
        Item 2
      </td>
      <td>
        Item 3
      </td>
    </tr>
  </thead>
</table>
`;

exports[`table matches snapshot 1`] = `
<table>
  <thead
    title="Header"
  >
    <tr
      title="Header"
    >
      <th>
        Head 1
      </th>
    </tr>
  </thead>
  <tbody
    title="Body Title"
  >
    <tr
      title="Body Row 1"
    >
      <td>
        Table Element 1
      </td>
    </tr>
  </tbody>
</table>
`;
